Invite flow refactoring : Part-2 of #949

kumarpalsinh25 opened the pull request:
This PR contains flow of invite users via super invite token of the space and chat.


Reference Tickets :
https://github.com/acterglobal/a3/issues/949

Reference Balsamiq Design :
https://balsamiq.cloud/sj2u1jx/plw5law/r18F4

References Video:

https://github.com/acterglobal/a3/assets/51526480/0e955221-ba54-4e51-b2e0-d105ce195dd4

gnunicorn commented on app/lib/features/invite_members/pages/invite_page.dart:
```diff
@@ -171,14 +201,131 @@ class InvitePage extends ConsumerWidget {
     );
   }
+
+  Widget _inviteCodeUI() {
```
I think there is enough within this widget and all the following functions allow to separate that out into its own widget file...

gnunicorn commented on app/lib/features/invite_members/pages/share_invite_code.dart:
```diff
@@ -0,0 +1,192 @@
+import 'package:go_router/go_router.dart';
+import 'package:share_plus/share_plus.dart';
+
+class ShareInviteCode extends ConsumerStatefulWidget {
+  final String inviteCode;
+
+  const ShareInviteCode({super.key, required this.inviteCode});
+
+  @override
+  ConsumerState<ConsumerStatefulWidget> createState() =>
+      _ShareInviteCodeState();
+}
+
+class _ShareInviteCodeState extends ConsumerState<ShareInviteCode> {
+  String userName = '';
+
+  @override
+  void initState() {
+    super.initState();
+    getUserName();
+  }
+
+  Future<void> getUserName() async {
+    final account = await ref.read(accountProfileProvider.future);
+    userName = account.profile.displayName ?? '';
+    setState(() {});
+  }
```
same uneccesary state pattern here. please use `ref.watch` instead.

gnunicorn commented on app/lib/features/super_invites/providers/super_invites_providers.dart:
```diff
@@ -31,8 +31,30 @@ final superInvitesForRoom = FutureProvider.autoDispose
+final inviteCodeForSelectRoomOnly = FutureProvider.autoDispose
+    .family<SuperInviteToken?, String>((ref, roomId) async {
+  // Get all tokens
+  final allInvites = await ref.watch(superInvitesTokensProvider.future);
+  // Get list of tokens with associate with given roomId
+  var allInvitesRelatedToRoomId = allInvites
+      .where(
+        (invite) =>
+            invite.rooms().map((e) => e.toDartString()).contains(roomId),
+      )
+      .toList();
```
this should reuse the existing provider rather than copy-paste the code ...
```suggestion
  final allInvitesRelatedToRoomId = await ref.watch(superInvietsForRoom(roomId).future);
```

kumarpalsinh25 replied:
Created additional provide to manage multiple invite code for specific room more precisely. Existing provide will return multiple tokens if available and on the other side, new provider which I created will return only single token or null for specific roomId. I will explain this on the call if it is confusing.

gnunicorn replied:
no, that's fine, if you look at my diff you'll see that I only asked for stuff to be reused...

kumarpalsinh25 replied:
Ah, understood. Noted!

gnunicorn commented on app/lib/features/super_invites/providers/super_invites_providers.dart:
```diff
@@ -31,8 +31,30 @@ final superInvitesForRoom = FutureProvider.autoDispose
+      )
+      .toList();
+  // Get single token which is associate with single roomId only
+  var inviteCodeWhichHaveSelectedRoomIdOnly = allInvitesRelatedToRoomId
```
```suggestion
  final inviteCodeWhichHaveSelectedRoomIdOnly = allInvitesRelatedToRoomId
```

gnunicorn reviewed (changes requested):
Video looks good (other than the dotted bordered icons, I am not a fan of those..) but the code is ... sup par. In particular the widget that are needlessly stateful and use a bad patter of `ref.read` need to be refactored before we want to merge them.

kumarpalsinh25 commented on app/ios/Runner/Info.plist:
```diff
@@ -54,5 +54,9 @@
+	<key>LSApplicationQueriesSchemes</key>
+    <array>
+        <string>whatsapp</string>
+    </array>
```
Haven't got specific solution for direct opening `signal` app. For time begin we can access it via generic share action button.

kumarpalsinh25 commented:
@gnunicorn 

1) Resolve PR Feedback with Below commit.
- https://github.com/acterglobal/a3/pull/1774/commits/4c5c5650c02cceb28f6f47baee2b3fe2a994fa57
- https://github.com/acterglobal/a3/pull/1774/commits/085415d439d23557dde0b7e6f92ab4f9e4a007b1

2) Changed the position of the invite button.
https://github.com/acterglobal/a3/pull/1774/commits/fcfbe23442891cbe462a361bd29790445c4d747c
![Screenshot 2024-05-20 at 4 35 52 PM](https://github.com/acterglobal/a3/assets/51526480/3d404e12-aa56-4f83-8b6e-11e079f82f4e)


3) Changed user invited button label and color.
https://github.com/acterglobal/a3/pull/1774/commits/df7dd712986cb895ecd860141200bd37ece78e8b
![Screenshot 2024-05-20 at 4 35 39 PM](https://github.com/acterglobal/a3/assets/51526480/5c9db23b-3c02-4497-8ecd-0c2dbf02d8fd)

gnunicorn reviewed (approved):
Just one remark about separating a widget, but not super urgent, could also be postponed

kumarpalsinh25 commented on app/lib/features/invite_members/pages/invite_page.dart:
```diff
@@ -171,14 +180,144 @@ class InvitePage extends ConsumerWidget {
     );
   }
+
+  Widget _inviteCodeUI(BuildContext context, WidgetRef ref) {
```
Intentionally haven't further separate code of `_inviteCodeUI` as there were already many smaller separate functions are in this file and code of `_inviteCodeUI` looks tightly couple. So better to leave it as it is.
